### PR TITLE
Cleanup `object_store::retry` client error handling

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Relates to #4880 

# Rationale for this change
 

The existing logic was hard to follow, with various different error variants overlayed onto the same structure. This makes these variants explicit, making the code easier to follow, and better able to support acting on error payloads.

# What changes are included in this PR?

# Are there any user-facing changes?

No, this error type is crate-private

### CODE REVIEW

File: object_store/src/client/retry.rs [hunk 1]
```diff
@@ -23,46 +23,50 @@ use futures::FutureExt;
...
-        self.status
+        match self {
+            Error::BareRedirect => None,
+            Error::Client { status, .. } => Some(*status),
+            Error::Response { source, .. } => source.status(),
+        }
+    }
+
+    /// Returns the error body if any
+    pub fn body(&self) -> Option<&str> {
```
Comment (tustvold) [inline object_store/src/client/retry.rs:60]: Being able to provide this method is what originally motivated this PR

File: object_store/src/client/retry.rs [hunk 2]
```diff
@@ -198,28 +199,31 @@ impl RetryExt for reqwest::RequestBuilder {
...
+                                        retries,
+                                        source: e,
+                                    }
+                                });
                             }
 
                             let sleep = backoff.next();
                             retries += 1;
-                            info!("Encountered server error, backing off for {} seconds, retry {} of {}", sleep.as_secs_f32(), retries, max_retries);
+                            info!("Encountered response error, backing off for {} seconds, retry {} of {}", sleep.as_secs_f32(), retries, max_retries);
```
Comment (tustvold) [inline object_store/src/client/retry.rs:226]: I thought this was less ambiguous, as this branch is for if we received an HTTP response with an error status

File: object_store/src/client/retry.rs [hunk 3]
```diff
@@ -238,16 +242,14 @@ impl RetryExt for reqwest::RequestBuilder {
...
-                                message: "request error".to_string(),
-                                status: e.status(),
-                                source: Some(e),
+                                source: e,
                             })
                         }
                         let sleep = backoff.next();
                         retries += 1;
-                        info!("Encountered request error ({}) backing off for {} seconds, retry {} of {}", e, sleep.as_secs_f32(), retries, max_retries);
+                        info!("Encountered transport error ({}) backing off for {} seconds, retry {} of {}", e, sleep.as_secs_f32(), retries, max_retries);
```
Comment (tustvold) [inline object_store/src/client/retry.rs:252]: I wasn't really sure what to call this, but this variant is for if the underlying connection/stream gets aborted without a response being received, so I went with transport to describe this

File: object_store/src/client/retry.rs [hunk 4]
```diff
@@ -23,46 +23,50 @@ use futures::FutureExt;
...
-impl std::error::Error for Error {
-    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
-        self.source.as_ref().map(|e| e as _)
-    }
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Received redirect without LOCATION, this normally indicates an incorrectly configured region"))]
+    BareRedirect,
+
+    #[snafu(display("Client error with status {status}: {}", body.as_deref().unwrap_or("No Body")))]
```
Comment (alamb) [inline object_store/src/client/retry.rs:36]: Isn't this techically a `ServerError` as is a value returned from the server (rather than some error trying to make the connecton?)

File: object_store/src/client/retry.rs [hunk 5]
```diff
@@ -23,46 +23,50 @@ use futures::FutureExt;
...
+    Client {
+        status: StatusCode,
+        body: Option<String>,
+    },
+
+    #[snafu(display("Response error after {retries} retries: {source}"))]
```
Comment (alamb) [inline object_store/src/client/retry.rs:42]: Isn't this kind of error normally called a `ClientError` or `ConnectionError` as it denotes a problem communicating with the server?

File: object_store/src/client/retry.rs (hunk 3 above)
Comment (alamb) [reply object_store/src/client/retry.rs:252]: I think transport error makes sense ✅

File: object_store/src/client/retry.rs (hunk 2 above)
Comment (alamb) [reply object_store/src/client/retry.rs:226]: I actually think `server error` is what I would expect. I would be confused about what a `response error` was

**Review by alamb — APPROVED** [review]
Thank you @tustvold 

I found the naming to be a little confusing, as I commented. 

I verified that this type does not appear to be public: https://docs.rs/object_store/0.7.1/object_store/?search=retry

Since it isn't a public API I think it could be merged as is (though I do think we should change the names / display of the errors)

File: object_store/src/client/retry.rs (hunk 5 above)
Comment (tustvold) [reply object_store/src/client/retry.rs]: I removed "Response" to avoid confusion

File: object_store/src/client/retry.rs [hunk 6]
```diff
@@ -39,8 +39,8 @@ pub enum Error {
         body: Option<String>,
     },
 
-    #[snafu(display("Response error after {retries} retries: {source}"))]
-    Response {
+    #[snafu(display("Error after {retries} retries: {source}"))]
```
Comment (alamb) [inline object_store/src/client/retry.rs:42]: 👍

**Review by alamb — APPROVED** [review]
